File: frontend mobile/lib/models/loan.dart
class LoanId {
  final int userId;
  final int bookId;

  LoanId({
    required this.userId,
    required this.bookId,
  });

  factory LoanId.fromJson(Map<String, dynamic> json) {
    return LoanId(
      userId: json['userId'],
      bookId: json['bookId'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'bookId': bookId,
    };
  }
}

class Loan {
  final LoanId id;
  final String? photo;
  final Map<String, dynamic> user;
  final Map<String, dynamic> book;
  final DateTime loanDate;
  final DateTime returnDate;
  final bool isReturned;

  Loan({
    required this.id,
    this.photo,
    required this.user,
    required this.book,
    required this.loanDate,
    required this.returnDate,
    required this.isReturned,
  });

  factory Loan.fromJson(Map<String, dynamic> json) {
    return Loan(
      id: LoanId.fromJson(json['id']),
      photo: json['photo'],
      user: json['user'],
      book: json['book'],
      loanDate: DateTime.parse(json['loanDate']),
      returnDate: DateTime.parse(json['returnDate']),
      isReturned: json['isReturned'] ?? false, // Default to false
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id.toJson(),
      if (photo != null) 'photo': photo,
      'user': user,
      'book': book,
      'loanDate': loanDate.toIso8601String(),
      'returnDate': returnDate.toIso8601String(),
      'isReturned': isReturned,
    };
  }
}

File: Backend/projet-multiplateforme-e2425g3_6-test/src/main/java/isme/testporjey/Models/Loan.java
package isme.testporjey.Models;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Loan {
    @EmbeddedId
    private LoanId id;
    private String photo;

    @ManyToOne
    @MapsId("userId")
    @JoinColumn(name = "user_id")
    private User user;

    @ManyToOne
    @MapsId("bookId")
    @JoinColumn(name = "book_id")
    private Book book;

    @NotNull
    private LocalDate loanDate;  // Correct field name
    @NotNull
    private LocalDate returnDate; // Correct field name
    @NotNull
    private boolean isReturned;

    @Override
    public String toString() {
        return "Loan(id=" + id +
                ", user=User{id=" + user.getId() + "}" +
                ", book=Book{id=" + book.getId() + ", title=" + book.getTitle() + ", photo=" + book.getPhoto() + ", author=" + book.getAuthor() + ", description=" + book.getDescription() + ", availableCopies=" + book.getAvailableCopies() + "}" +
                ", loanDate=" + loanDate +
                ", returnDate=" + returnDate +
                ", isReturned=" + isReturned + ")";
    }
}
